ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs

```diff
@@ -0,0 +1,715 @@
+  Trace.traceWith dbTracer TDBOpening
+
+  dbOpenHandles <- IOLike.newTVarIO Map.empty
+  dbClosed      <- IOLike.newTVarIO False
+
+  let
+    path = FS.mkFsPath ["tables"]
+
+  -- get the filepath for this db creates the directory if appropriate
+  dbFilePath <- guardDbDirWithRetry GDDMustNotExist sfs path
+
+  -- copy from another lmdb path if appropriate
+  case initFrom of
+    HD.InitFromCopy (HD.BackingStorePath fp) -> initFromLMDBs dbTracer limits sfs fp path
+    _                                        -> pure ()
+
+  -- open this database
+  dbEnv <- liftIO $ LMDB.openEnvironment dbFilePath (unLMDBLimits limits)
+
+  -- Create the settings table.
+  dbSettings <- liftIO $ LMDB.readWriteTransaction dbEnv $ LMDB.getDatabase (Just "_dbsettings")
+  liftIO $ LMDB.readWriteTransaction dbEnv $ writeDbSettings dbSettings
+
+  -- The LMDB.Database that holds the @`DbState`@ (i.e. sequence number)
+  -- This transaction must be read-write because on initialisation it creates the database
+  dbState <- liftIO $ LMDB.readWriteTransaction dbEnv $ LMDB.getDatabase (Just "_dbstate")
+
+  -- Here we get the LMDB.Databases for the tables of the ledger state
+  -- Must be read-write transaction because tables may need to be created
+  dbBackingTables <- liftIO $ LMDB.readWriteTransaction dbEnv $
+    traverseLedgerTables getDb namesLedgerTables
+
+  dbNextId <- IOLike.newTVarIO 0
+
+  let
+    db = Db { dbEnv
+            , dbSettings
+            , dbState
+            , dbBackingTables
+            , dbFilePath
+            , dbTracer
+            , dbClosed
+            , dbOpenHandles
+            , dbNextId
+            }
+
+    bsClose :: m ()
+    bsClose = do
+      guardDbClosed dbClosed
+      Trace.traceWith dbTracer $ TDBClosing dbFilePath
+      openHandles <- IOLike.readTVarIO dbOpenHandles
+      for_ openHandles HD.bsvhClose
+      IOLike.atomically $ IOLike.modifyTVar dbClosed (const True)
+      liftIO $ LMDB.closeEnvironment dbEnv
+      Trace.traceWith dbTracer $ TDBClosed dbFilePath
+
+    bsCopy shfs (HD.BackingStorePath to0) = do
+      guardDbClosed dbClosed
+      to <- guardDbDir GDDMustNotExist shfs to0
+      lmdbCopy dbTracer dbEnv to
+
+    bsValueHandle =
+      guardDbClosed dbClosed >>
+      mkLMDBBackingStoreValueHandle db
+
+    bsWrite :: SlotNo -> LedgerTables l DiffMK -> m ()
+    bsWrite slot diffs = do
+      guardDbClosed dbClosed
+      oldSlot <- liftIO $ LMDB.readWriteTransaction dbEnv $ withDbStateRW dbState $ \s@DbState{dbsSeq} -> do
+        unless (dbsSeq <= At slot) $ liftIO . throwIO $ DbErrNonMonotonicSeq (At slot) dbsSeq
+        void $ zipLedgerTables2A writeLMDBTable dbBackingTables codecLedgerTables diffs
+        pure (dbsSeq, s {dbsSeq = At slot})
+      Trace.traceWith dbTracer $ TDBWrite oldSlot slot
+
+  -- now initialise those tables if appropriate
+  case initFrom of
+      HD.InitFromValues slot vals -> initFromVals dbTracer slot vals db
```

Comment:
This flow is a bit obscure, probably could be split in 2 steps: getting the db and maybe populating it